### ISSUE TITLE
Pkg operations should be case-sensitive on case-insensitive filesystems

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#17747]
e.g. I shouldn't be able to do `Pkg.test("pycall")`.  Since packages are module names, and module loading is case-sensitive, Pkg should be too.

I just noticed this because `Pkg.test("NBinclude")` failed for a subtle reason (the tests loaded, but with an altered path) and it took me a while to realize that it was because I had mistyped `Pkg.test("NBInclude")`.

**Comment by stevengj** [comment]
(Should be able to use the same machinery as #13542.)

**Comment by stevengj** [comment]
cc @malmaud

**Comment by jakebolewski** [comment]
I've fixed this, PR forthcoming once I figure out how to add tests.

**Comment by stevengj** [comment]
@jakebolewski, I also have a patch forthcoming.  Can we compare notes?

**Comment by jakebolewski** [comment]
Sure, I didn't do anything special.  Just made sure all the path relevant path comparisons in Pkg were case sensitive to the package name.  @malmaud's  code for OSX works for any inodes if you strip out trailing `/` for directories.

**Comment by stevengj** [comment]
See the commit linked above.   I moved the `isfoo_casesensitive` checks to `stat.jl`.  The tricky part was finding the relevant entry points in `Pkg`.

I didn't think to strip out a trailing `/`, though.

**Comment by jakebolewski** [comment]
Well the two are pretty much identical sans stripping out trailing `/`, and yours has docs 👍

**Comment by jakebolewski** [comment]
I would advocate that case normalization becomes apart of the `isfile`, `isdir`, `ispath` API through a keyword argument.  But that seems like an API change for 0.6.

**Comment by stevengj** [comment]
Before we can make it part of the public API, we need to deal with the issue that currently the case is only checked for `basename(path)` on MacOS.

**Comment by stevengj** [comment]
Do you want to pull my patch, add your trailing `/` fix and any other differences, then submit a PR?  (Or merge in the other direction if you prefer.)

**Comment by stevengj** [comment]
(Anyway, the change to the public `isfoo` API seems like it should go in a separate PR after the Pkg fix.)

**Comment by jakebolewski** [comment]
Sure, I can add to that commit.

**Comment by simonbyrne** [comment]
This is now fixed.
